### PR TITLE
 Picard liftover vcf

### LINKED ISSUE #1402: new module: Picard/LiftoverVcf
I think it would be good to have a module for [Picard/LiftoverVcf](https://bioconda.github.io/recipes/picard/README.html)

- [x] This module does not exist yet with the [`nf-core modules list`](https://github.com/nf-core/tools#list-modules) command
- [x] There is no [open pull request](https://github.com/nf-core/modules/pulls) for this module
- [x] There is no [open issue](https://github.com/nf-core/modules/issues) for this module
- [x] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by Lucpen**
## PR checklist

Closes #1402 

- [x] Created a module for Picard liftover vcf.
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/picard/liftovervcf/meta.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      type: file
+      description: File containing software versions
+      pattern: "versions.yml"
+  - vcf:
+      type: file
+      description: VCF file containing successfully lifted variants
+      pattern: "*.{vcf,vcf.gz}"
+  - vcf_unlifted:
+      type: file
+      description: VCF file containing unsuccessfully lifted variants
```
Comment (Gwennid) [inline modules/picard/liftovervcf/meta.yml:48]: Alternative description: "VCF files containing variants that couldn't be lifted."

Should the file pattern be `".{unlifted.vcf,unlifted.vcf.gz}"` ? (if yes, do the same for the other VCF output)

File: modules/picard/liftovervcf/main.nf [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    path "versions.yml"                   , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def avail_mem = 1
+
```
[the remark this answers is not in the document]
Comment (sateeshperi) [reply modules/picard/liftovervcf/main.nf:29]: this setting has been used in other picard modules. plz consider adding if you think its ok

File: tests/modules/picard/liftovervcf/test.yml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+- name: picard liftovervcf test_picard_liftovervcf
+  command: nextflow run tests/modules/picard/liftovervcf -entry test_picard_liftovervcf -c tests/config/nextflow.config
+  tags:
+    - picard
+    - picard/liftovervcf
+  files:
+    - path: output/picard/test.lifted.vcf
+      contains:
+        - "chr22"
```
Comment (Gwennid) [inline tests/modules/picard/liftovervcf/test.yml:9]: Do you know why the VCF containing the lifted variants does not always have the same md5sum?

File: modules/picard/liftovervcf/main.nf [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        'quay.io/biocontainers/picard:2.26.10--hdfd78af_0' }"
+
+    input:
+    tuple val(meta), path(input_vcf)
+    path dict
+    path chain
+    path fasta
+
+    output:
+    tuple val(meta), path("*lifted.vcf")  , emit: lifted
```
Comment (Gwennid) [inline modules/picard/liftovervcf/main.nf:18]: It is recommended to have a compressed file as output. Is that possible here?

**Comment by Gwennid** [comment]
Hi Lucia,
I think this looks good! I just had a few questions/comments. I am trying to run the test on our system and haven't succeeded so far, but I think it's due to our system not to your code :) I'll try some more.

File: modules/picard/liftovervcf/meta.yml (hunk 1 above)
Comment (Lucpen) [reply modules/picard/liftovervcf/meta.yml:48]: Thanks for the suggestion, I have added it in the latest commit

File: tests/modules/picard/liftovervcf/test.yml (hunk 3 above)
Comment (Lucpen) [reply tests/modules/picard/liftovervcf/test.yml:9]: The lifted.vcf.gz contains the path for the reference, which changes every time the test is run so the md5sum changes too. You can see this on the line of the lifted.vcf starting with ##reference=
It is not the case for the unlifted.vcf, so its md5sum is stable

File: modules/picard/liftovervcf/main.nf (hunk 4 above)
Comment (Lucpen) [reply modules/picard/liftovervcf/main.nf]: Fixed :)

File: tests/modules/picard/liftovervcf/test.yml (hunk 3 above)
Comment (Gwennid) [reply tests/modules/picard/liftovervcf/test.yml:9]: Ok!

File: modules/picard/liftovervcf/meta.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  - meta:
+      type: map
+      description: Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - input_vcf:
+      type: file
+      description: VCF file
+      pattern: "*.{vcf,vcf.gz}"
+  - chain:
+      type: file
```
Comment (sateeshperi) [inline modules/picard/liftovervcf/meta.yml:25]:
```suggestion
      type: file
      description: The liftover chain file
```

File: modules/picard/liftovervcf/meta.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+  - fasta:
+      type: file
+      description: fasta file
```
Comment (sateeshperi) [inline modules/picard/liftovervcf/meta.yml:28]:
```suggestion
      description: fasta file
      pattern: "*.fasta"
```

File: modules/picard/liftovervcf/meta.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+  - dict:
+      type: file
+      description: dictionary for fasta file
```
Comment (sateeshperi) [inline modules/picard/liftovervcf/meta.yml:31]:
```suggestion
      description: dictionary for fasta file
      pattern: "*.{dict}"
```

**Review by sateeshperi — APPROVED** [review]
lgtm! Thanks for working with the reviews

File: modules/picard/liftovervcf/main.nf [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        log.info '[Picard LiftoverVcf] Available memory not known - defaulting to 1GB. Specify process memory requirements to change this.'
+    } else {
+        avail_mem = task.memory.giga
+    }
+    """
+    picard \\
+        -Xmx${avail_mem}g \\
+        LiftoverVcf \\
+        $args \\
+        I=$input_vcf \\
```
Comment (FriederikeHanssen) [inline modules/picard/liftovervcf/main.nf:38]: @maxulysse what did you decide now for parameters during the picard/gatk refactoring?

File: tests/modules/picard/liftovervcf/test.yml [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    - picard/liftovervcf
+    - picard
+  files:
+    - path: output/picard/test.lifted.vcf.gz
+      contains:
+        - "chr22"
+    - path: output/picard/test.unlifted.vcf.gz
+      md5sum: 43d760d50cb49e461c9fc3949fa4fc03
+    - path: output/picard/versions.yml
+      md5sum: b16383e5111c1850cb2c8f97353001d0
```
Comment (jfy133) [inline tests/modules/picard/liftovervcf/test.yml:13]: MD5sum Can be removed here to make it easier to update the module in the future with version bumps,

**Review by jfy133 — APPROVED** [review]
LGTM

File: tests/modules/picard/liftovervcf/test.yml (hunk 9 above)
Comment (sateeshperi) [reply tests/modules/picard/liftovervcf/test.yml:13]: @Lucpen

Comment (Lucpen) [reply tests/modules/picard/liftovervcf/test.yml:13]: @jfy133 should I simply delete the md5sum or replace them with something else like contains?

File: modules/picard/liftovervcf/main.nf (hunk 8 above)
Comment (sateeshperi) [reply modules/picard/liftovervcf/main.nf:38]: ping @maxulysse

File: tests/modules/picard/liftovervcf/test.yml (hunk 9 above)
Comment (jfy133) [reply tests/modules/picard/liftovervcf/test.yml]: You can just delete, not a problem. As long as the file is there it should have worked :+1:

**Review by maxulysse — APPROVED** [review]

**Pull request merged**